include/miral/miral/wayland_extensions.h

```diff
@@ -119,6 +123,20 @@ class WaylandExtensions
     struct Self;
     std::shared_ptr<Self> self;
 };
+
+/// Get the MirAL application for a wl_client.
+/// \return The application (null if no application is found)
+/// \remark Since MirAL 2.5
+auto application_for(wl_client* client) -> Application;
+
+/// Get the MirAL Window for a Wayland Surface, XdgSurface, etc.
+/// Note that there may not be a corresponding miral::Window (e.g. the
+/// surface is created and assigned properties before 'commit' creates the
+/// miral::Window).
+///
+/// \return The window (null if no window is found)
+/// \remark Since MirAL 2.5
+auto window_for(wl_client* client, wl_resource* surface) -> Window;
```

Comment:
1. Is some sort of optional a thing we can return? Would do a better job signalling it can be null, and forcing those using this API into dealing with that case

2. `wl_resource_get_client()` is a thing that exists, so why require the client get sent in separately (assuming the client must always be the one the surface resource is on)

Reply:
I wonder whether this wouldn't be better as
`miral::dynamic_cast<Target>(wl_resource *) -> std::experimental::optional<Target>` or moral equivalent.

As @wmww says, we don't need the `wl_client*`, as each `wl_resource` is already associated with its client.

We'll have a bunch of these accessors, and they look quite a lot like downcasts (“I know this `wl_resource` is really a `Window`, please give me the `Window`). It seems like it might be convenient to consolidate them into a downcast-like template?

Reply:
(This doesn't apply to `miral::application_for(wl_client*)`, as we'll only map a single MirAL object to a `wl_client` (?))

Reply:
1. `Window` already has `operator bool()` - doesn't that make adding an optional/fallible return type redundant?

2. That makes sense. I was unconsciously reluctant to introduce a direct dependency of libmiral on libwayland.

3. I'm not convinced of the need for a wrapper template. (The logic cannot be implemented directly in the template without exposing unstable ABIs.) But we _could_ add it later - simply add a specialization that forwards to `window_for()`.